File: android/designsystem/src/main/kotlin/com/bolsillo/designsystem/component/SegmentedControl.kt
package com.bolsillo.designsystem.component

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.bolsillo.designsystem.theme.BolsilloTheme

@Composable
fun <T> SegmentedControl(
    options: List<T>,
    selected: T,
    label: (T) -> String,
    modifier: Modifier = Modifier,
    onSelected: (T) -> Unit,
) {
    Row(
        modifier =
            modifier
                .fillMaxWidth()
                .clip(BolsilloTheme.shapes.control)
                .background(BolsilloTheme.colors.track)
                .padding(4.dp),
        horizontalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        options.forEach { option ->
            val isSelected = option == selected
            Box(
                modifier =
                    Modifier
                        .weight(1f)
                        .clip(BolsilloTheme.shapes.control)
                        .let { m ->
                            if (isSelected) {
                                m.shadow(BolsilloTheme.elevation.segmentedThumb, BolsilloTheme.shapes.control)
                                    .background(BolsilloTheme.colors.surface)
                            } else {
                                m
                            }
                        }
                        .clickable(role = Role.Tab) { onSelected(option) }
                        .semantics {
                            this.role = Role.Tab
                            this.selected = isSelected
                        }
                        .padding(vertical = 10.dp),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    text = label(option),
                    color = if (isSelected) BolsilloTheme.colors.textPrimary else BolsilloTheme.colors.textMuted,
                    style = BolsilloTheme.typography.label,
                )
            }
        }
    }
}

File: ios/Packages/Bolsillo/Sources/BolsilloDesignSystem/Components/SegmentedControl.swift
import SwiftUI

/// Custom segmented control that matches Bolsillo's visual language.
/// Renders a pill-thumb sliding over a track, with animated selection.
public struct SegmentedControl<T: Hashable>: View {
    public let options: [(label: String, value: T)]
    @Binding public var selection: T

    @Environment(\.bolsilloTheme) private var theme
    @Namespace private var thumbNamespace

    public init(options: [(label: String, value: T)], selection: Binding<T>) {
        self.options    = options
        self._selection = selection
    }

    public var body: some View {
        HStack(spacing: 0) {
            ForEach(options.indices, id: \.self) { index in
                let option = options[index]
                let isSelected = option.value == selection

                Button {
                    withAnimation(.spring(response: 0.3, dampingFraction: 0.75)) {
                        selection = option.value
                    }
                } label: {
                    Text(option.label)
                        .font(isSelected ? theme.typography.bodyStrong : theme.typography.body)
                        .foregroundStyle(isSelected ? theme.colors.textPrimary : theme.colors.textMuted)
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, theme.spacing.sm)
                        .background {
                            if isSelected {
                                RoundedRectangle(cornerRadius: theme.radius.control, style: .continuous)
                                    .fill(theme.colors.surface)
                                    .elevate(theme.elevation.segmentedThumb)
                                    .matchedGeometryEffect(id: "thumb", in: thumbNamespace)
                            }
                        }
                }
                .buttonStyle(.plain)
            }
        }
        .padding(theme.spacing.xs)
        .background(
            RoundedRectangle(cornerRadius: theme.radius.control + theme.spacing.xs, style: .continuous)
                .fill(theme.colors.track)
        )
    }
}
